### PR TITLE
Add RST documentation

### PR DESCRIPTION
**Pull request by msricher**

### CODE REVIEW

File: doc/gauss_newton.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    : MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    : GNU General Public License for more details.
+    :
+    : You should have received a copy of the GNU General Public License
+    : along with this program; if not, see <http://www.gnu.org/licenses/>
+
+Gauss–Newton Algorithm
+======================
+
+The advantage of Gauss-Newton Method is that the second derivatives are not necessarily needed to be
```
Comment (matt-chan) [inline doc/gauss_newton.rst]: do not necessarily need

File: doc/gauss_newton.rst [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+computed.The limitation is also obvious since it can only handle with the minimization the sum of
```
Comment (matt-chan) [inline doc/gauss_newton.rst]: - space before The. 
- "since it can only minimize the sum of squared function values".

File: doc/gauss_newton.rst [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+squared function values. The Gaussian-Newton is mainly used to resolve the non-linear least squares
```
Comment (matt-chan) [inline doc/gauss_newton.rst]: Gauss-Newton method

File: doc/gauss_newton.rst [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+problems, named after two famous mathematicians Carl Friedrich Gauss and Isaac Newton.
```
Comment (matt-chan) [inline doc/gauss_newton.rst]: problems. It is named after the two

File: doc/gauss_newton.rst [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+.. math::
+    (J^{\top} J)(x_{s+1} - x_s) = -J^{\top} f
+
+where :math:`J` is the Jacobian matrix, :math:`J^{\top}` is the transpose of Jacobian, :math:`s` and
+:math:`s+1` are the indexes. Now the problem becomes
+
+.. math::
+    x_{s+1} = x_s - {(J^{\top} J)}^{-1} {J^{\top}} f(x_s)
+
+where an initial guess :math:`x_0`. When the matrix :math:`f` is a squared one, the simplified
```
Comment (matt-chan) [inline doc/gauss_newton.rst]: - where :math:`x_0` is the initial guess.
- is squared

File: doc/jacobian.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    : You should have received a copy of the GNU General Public License
+    : along with this program; if not, see <http://www.gnu.org/licenses/>
+
+Jacobian Matrix
+===============
+
+Definition
+----------
+
+Let's suppose we have a vector valued function of a vector
```
Comment (matt-chan) [inline doc/jacobian.rst]: - Delete Let's
- Delete of a vector

File: doc/jacobian.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+:math:`f: \mathbb{R}^n \rightarrow \mathbb{R}^m`. This function will take the input vector
+:math:`x \in \mathbb{R}^n` and maps it to
```
Comment (matt-chan) [inline doc/jacobian.rst]: map

File: doc/jacobian.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+the result vector :math:`f(x) \in \mathbb{R}^m`. The Jacobian is can be also termed as
```
Comment (matt-chan) [inline doc/jacobian.rst]: The Jacobian is also called

File: doc/jacobian.rst [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+         \right]
+    & = {
+    \begin{bmatrix}
+      \frac{\partial f_1}{\partial x_1} & \cdots & \frac{\partial f_1}{\partial x_n} \\
+      \vdots         & \ddots & \vdots \\
+      \frac{\partial f_m}{\partial x_1} & \cdots & \frac{\partial f_m}{\partial x_n}
+    \end{bmatrix}
+    }
+
+Or put it in a element-wise way,
```
Comment (matt-chan) [inline doc/jacobian.rst]: Or, element-wise,

File: doc/jacobian.rst [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                  2xy & x^2 \\
+                  5   & \cos y
+                \end{bmatrix}
+               }
+
+Solver
+------
+
+In our package, we have two different ways to approximate Jacobian, namely the forward difference
+method and the central difference one. All these two methods are based on the principle that the
```
Comment (matt-chan) [inline doc/jacobian.rst]: - s/one/method/
- s/All/both/
- s/the//

File: doc/newton.rst [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    : MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    : GNU General Public License for more details.
+    :
+    : You should have received a copy of the GNU General Public License
+    : along with this program; if not, see <http://www.gnu.org/licenses/>
+
+Newton Algorithm
+================
+
+Newton method (also knows as the Newton-Raphson method, Newton's iteration) is a algorithm that utilizes the first terms of Taylor Polynomials of function :math:`f(x)` to find the roots of
```
Comment (matt-chan) [inline doc/newton.rst]: - Newton's method
- s/a/an/
- s/of Taylor Polynomials/the taylor series/

File: doc/newton.rst [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+:math:`f(x)=0`. For the Taylor series of :math:`f(x)` about the pint
```
Comment (matt-chan) [inline doc/newton.rst]: point

File: doc/newton.rst [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+:math:`x_{s+1} = x_s + \epsilon` where the :math:`s` and :math:`s+1` are indexes,
```
Comment (matt-chan) [inline doc/newton.rst]: s/the//

File: doc/newton.rst [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+.. math::
+    f(x_{s+1}) = f(x_s) + f^{\prime}(x_s) \epsilon + \frac{1}{2} f^{\prime \prime}(x_s) + \cdots
+
+If we only keep the first term, we will get
+
+.. math::
+    f(x_{s+1}) \approx f(x_s) + f^{\prime}({x_s}) \epsilon
+
+Because we are solving :math:`f(x)=0`, then for Newton method, the problem can be formulated as
```
Comment (matt-chan) [inline doc/newton.rst]: Newton's method

**Review by matt-chan — CHANGES_REQUESTED** [review]
Grammar fixes needed. I didn't check the math because github doesn't render it for some reason.

File: doc/gauss_newton.rst (hunk 3 above)
Comment (fwmeng88) [reply doc/gauss_newton.rst]: Fixed.

File: doc/gauss_newton.rst (hunk 4 above)
Comment (fwmeng88) [reply doc/gauss_newton.rst]: Fixed.

File: doc/gauss_newton.rst (hunk 5 above)
Comment (fwmeng88) [reply doc/gauss_newton.rst]: Fixed.

File: doc/gauss_newton.rst (hunk 2 above)
Comment (fwmeng88) [reply doc/gauss_newton.rst]: 1. Why is space needed? When I compile the website, it looks like something unusual.
2. Fixed.

File: doc/jacobian.rst (hunk 6 above)
Comment (fwmeng88) [reply doc/jacobian.rst]: Done.

File: doc/jacobian.rst (hunk 7 above)
Comment (fwmeng88) [reply doc/jacobian.rst]: Done.

File: doc/jacobian.rst (hunk 8 above)
Comment (fwmeng88) [reply doc/jacobian.rst]: Done

File: doc/jacobian.rst (hunk 9 above)
Comment (fwmeng88) [reply doc/jacobian.rst]: Done

File: doc/newton.rst (hunk 11 above)
Comment (fwmeng88) [reply doc/newton.rst]: Done

File: doc/newton.rst (hunk 12 above)
Comment (fwmeng88) [reply doc/newton.rst]: Done.

File: doc/newton.rst (hunk 13 above)
Comment (fwmeng88) [reply doc/newton.rst]: done

File: doc/newton.rst (hunk 14 above)
Comment (fwmeng88) [reply doc/newton.rst]: Done

**Review by matt-chan — APPROVED** [review]

**Comment by msricher** [comment]
Closed; branch was messed up

**Pull request closed without merging**